### DISCUSSION TITLE
What is different for onStateChange and onGatheringStateChange

### DISCUSSION THREAD

**Discussion by lokenetwork** [paullouisageneau/libdatachannel#1056] [Q&A]
What is different for PeerConnection::onStateChange and PeerConnection::onGatheringStateChange.

When this function start run, I want to know when they run?

Thank you.

**Comment by voluntas** [comment]
PeerConnection::onStateChange is equivalent to connectionstatechange. <del>However, unlike connectionstatechange, onStateChange does not look at DTLS behavior. Please be careful about that.</del>

https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/connectionstatechange_event

PeerConnection::onGatheringStateChange is equivalent to icegatheringstatechange.

https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/icegatheringstatechange_event

It is based on W3C's WebRTC API, so it is recommended to read the source code while comparing it with W3C's WebRTC API.

https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection

**Reply by lokenetwork** [reply]
Thank you.

**Comment by paullouisageneau** [comment]
> However, unlike connectionstatechange, onStateChange does not look at DTLS behavior. Please be careful about that.

Actually, `onStateChange` looks at DTLS state. The difference with `connectionstatechange` in the reference WebRTC API is that if there are data channels, `Connected` will be reached only when SCTP is connected too.

**Reply by voluntas** [reply]
Oops, my bad, I couldn't find State in DTLS so I assumed you hadn't seen it.

**Reply by lokenetwork** [reply]
Thank you
